chain/src/chain.rs

```diff
@@ -1247,16 +1249,33 @@ fn setup_head(
 			let tip = Tip::from_header(&genesis.header);
 			batch.save_head(&tip)?;
 
-			// Initialize our header MM with the genesis header.
-			txhashset::header_extending(txhashset, &mut batch, |extension| {
-				extension.apply_header(&genesis.header)?;
-				Ok(())
-			})?;
-
 			batch.save_block_header(&genesis.header)?;
 
+			if genesis.kernels().len() > 0 {
+				let (utxo_sum, kernel_sum) = (sums, &genesis as &Committed).verify_kernel_sums(
+					genesis.header.overage(),
+					genesis.header.total_kernel_offset(),
+				)?;
+				sums = BlockSums {
+					utxo_sum,
+					kernel_sum,
+				};
+				txhashset::extending(txhashset, &mut batch, |extension| {
+					extension.apply_block(&genesis)?;
+					extension.validate_roots()?;
+					extension.validate_sizes()?;
+					Ok(())
+				})?;
+			} else {
+				// Initialize our header MMR with the genesis header.
+				txhashset::header_extending(txhashset, &mut batch, |extension| {
```

Comment:
Could we just `apply_block()` in both cases (genesis empty or non-empty) and have `validate_roots()` and `validate_sizes()` do the "right thing" for an empty genesis block?

Reply:
Also great idea, looks like using `apply_block()` just works. The 2 validation methods do nothing for the genesis block and I think that's fair for now. This way we can use the root to embed a hash.